src/library/features/crates/cratemanager.cpp

```diff
@@ -0,0 +1,293 @@
+            "WHERE %2 NOT IN (0,1)").arg(
+              CRATE_TABLE,
+              CRATETABLE_AUTODJ_SOURCE));
+        if (query.execPrepared() && (query.numRowsAffected() > 0)) {
+            kLogger.warning()
+                    << "Fixed boolean values in table" << CRATE_TABLE
+                    << "column" << CRATETABLE_AUTODJ_SOURCE
+                    << "for" << query.numRowsAffected() << "crates";
+        }
+    }
+
+    // Crate tracks
+    {
+        // Remove tracks from non-existent crates
+        FwdSqlQuery query(
+          database, QString(
+            "DELETE FROM %1 "
+            "WHERE %2 NOT IN "
+            "(SELECT %3 FROM %4)").arg(
+              CRATE_TRACKS_TABLE,
+              CRATETRACKSTABLE_CRATEID,
+              CRATETABLE_ID,
+              CRATE_TABLE));
+        if (query.execPrepared() && (query.numRowsAffected() > 0)) {
+            kLogger.warning()
+                    << "Removed" << query.numRowsAffected()
+                    << "crate tracks from non-existent crates";
+        }
+    }
+    {
+        // Remove library purged tracks from crates
+        FwdSqlQuery query(
+          database, QString(
+            "DELETE FROM %1 "
+            "WHERE %2 NOT IN "
+            "(SELECT %3 FROM %4)").arg(
+              CRATE_TRACKS_TABLE,
+              CRATETRACKSTABLE_TRACKID,
+              LIBRARYTABLE_ID,
+              LIBRARY_TABLE));
+        if (query.execPrepared() && (query.numRowsAffected() > 0)) {
+            kLogger.warning()
+                    << "Removed" << query.numRowsAffected()
+                    << "library purged tracks from crates";
+        }
+    }
+
+    // Crate hierarchy
+    {
+        // Remove paths from non-existent crates
+        FwdSqlQuery query(
+          database, QString(
+            "DELETE FROM %1 "
+            "WHERE %2 NOT IN "
+            "(SELECT %3 FROM $4)").arg(
+              CRATE_PATH_TABLE,
+              PATHTABLE_CRATEID,
+              CRATE_TABLE,
+              CRATETABLE_ID));
+        if (query.execPrepared() && (query.numRowsAffected() > 0)) {
+            kLogger.warning()
+                << "Removed" << query.numRowsAffected()
+                << "paths from non-existent crates";
+        }
+    }
+}
+
+void CrateManager::connectDatabase(QSqlDatabase database) {
+    m_database = database;
+
+    m_crateStorage.initialize(database);
+    m_crateTracks.initialize(database);
+    m_crateHierarchy.initialize(database);
+
+    createViews();
+}
+
+void CrateManager::disconnectDatabase() {
+    // Ensure that we don't use the current database connection
+    // any longer.
+    m_database = QSqlDatabase();
+}
+
+bool CrateManager::onPurgingTracks(const QList<TrackId> &trackIds) {
+    return m_crateTracks.onPurgingTracks(trackIds);
+}
+
+void CrateManager::checkClosure() {
+    if (m_crateHierarchy.countCratesInClosure() != m_crateStorage.countCrates()) {
+        m_crateHierarchy.reset(&storage());
```

Comment:
You should at least log a warning if you detect that the crate tables have become inconsistent.